From 54f629a6ce7be52de73c740f5b41d5821204b38e Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:49:54 -0400
Subject: [tools] lib/traceevent: Report better error message on bad function args

Message-id: <1408953210-26343-107-git-send-email-jolsa@redhat.com>
Patchwork-id: 88907
O-Subject: [PATCH RHEL7.1 BZ1133083 106/322] tools lib traceevent: Report better error message on bad function args
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 3a3ffa2e82205921d1189f1055c22dae4c72819a
Author: Steven Rostedt <rostedt@goodmis.org>
Date: Mon Nov 18 21:38:20 2013 -0500

description
===========
When Jiri Olsa was writing a function callback for
scsi_trace_parse_cdb(), he thought that the traceevent library had a
bug in it because he was getting this error:

  Error: expected ')' but read ','
  Error: expected ')' but read ','
  Error: expected ')' but read ','
  Error: expected ')' but read ','

But in truth, he didn't have the write number of arguments for the
function callback, and the error was the library detecting the
discrepancy. A better error message would have prevented the confusion:

  Error: function 'scsi_trace_parse_cdb()' only expects 2 arguments but event scsi_dispatch_cmd_timeout has more
  Error: function 'scsi_trace_parse_cdb()' only expects 2 arguments but event scsi_dispatch_cmd_start has more
  Error: function 'scsi_trace_parse_cdb()' only expects 2 arguments but event scsi_dispatch_cmd_error has more
  Error: function 'scsi_trace_parse_cdb()' only expects 2 arguments but event scsi_dispatch_cmd_done has more

Or

  Error: function 'scsi_trace_parse_cdb()' expects 4 arguments but event scsi_dispatch_cmd_timeout only uses 3
  Error: function 'scsi_trace_parse_cdb()' expects 4 arguments but event scsi_dispatch_cmd_start only uses 3
  Error: function 'scsi_trace_parse_cdb()' expects 4 arguments but event scsi_dispatch_cmd_error only uses 3
  Error: function 'scsi_trace_parse_cdb()' expects 4 arguments but event scsi_dispatch_cmd_done only uses 3
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/lib/traceevent/event-parse.c b/tools/lib/traceevent/event-parse.c
index 9849873..22566c2 100644
--- a/tools/lib/traceevent/event-parse.c
+++ b/tools/lib/traceevent/event-parse.c
@@ -2710,7 +2710,6 @@ process_func_handler(struct event_format *event, struct pevent_function_handler
  struct print_arg *farg;
  enum event_type type;
  char *token;
- const char *test;
  int i;
 
  arg->type = PRINT_FUNC;
@@ -2727,15 +2726,19 @@ process_func_handler(struct event_format *event, struct pevent_function_handler
   }
 
   type = process_arg(event, farg, &token);
-  if (i < (func->nr_args - 1))
-   test = ",";
-  else
-   test = ")";
-
-  if (test_type_token(type, token, EVENT_DELIM, test)) {
-   free_arg(farg);
-   free_token(token);
-   return EVENT_ERROR;
+  if (i < (func->nr_args - 1)) {
+   if (type != EVENT_DELIM || strcmp(token, ",") != 0) {
+    warning("Error: function '%s()' expects %d arguments but event %s only uses %d",
+     func->name, func->nr_args,
+     event->name, i + 1);
+    goto err;
+   }
+  } else {
+   if (type != EVENT_DELIM || strcmp(token, ")") != 0) {
+    warning("Error: function '%s()' only expects %d arguments but event %s has more",
+     func->name, func->nr_args, event->name);
+    goto err;
+   }
   }
 
   *next_arg = farg;
@@ -2747,6 +2750,11 @@ process_func_handler(struct event_format *event, struct pevent_function_handler
  *tok = token;
 
  return type;
+
+err:
+ free_arg(farg);
+ free_token(token);
+ return EVENT_ERROR;
 }
 
 static enum event_type
-- 
1.7.1